Migrate JMX connector to new APIs

takezoe opened the pull request.

findepi commented on presto-jmx/src/main/java/io/prestosql/plugin/jmx/JmxMetadata.java:
```diff
@@ -19,18 +19,8 @@
-import io.prestosql.spi.connector.Constraint;
-import io.prestosql.spi.connector.SchemaTableName;
-import io.prestosql.spi.connector.SchemaTablePrefix;
+import io.prestosql.spi.connector.*;
```
We do not use star imports.
Hopefully importing https://github.com/airlift/codestyle into Intellij should get you the right configuration.

takezoe replied:
Thanks! Fixed it.

findepi commented on presto-jmx/src/main/java/io/prestosql/plugin/jmx/JmxMetadata.java:
```diff
@@ -96,6 +87,22 @@ public JmxTableHandle getTableHandle(ConnectorSession session, SchemaTableName t
+            Optional.empty(),
+            Optional.empty(),
+            Optional.empty(),
+            emptyList());
```
2 formatting nits (somehow this comment got lost when i first wrote it)

- per code style, when each param is on a separate line, there should be a newline before the first one:
  ```
   return new ConnectorTableProperties(
       TupleDomain.all(),
   ```
- by convention, we use `ImmutableList.of()` rather than `emptyList()` (i understand this is subjective, this is just the convention we try to adhere to, as we maintain codebase consistency)

takezoe replied:
Following the convention of project makes sense. I fixed them.

takezoe commented:
@findepi Thanks for your review! I fixed code and updated the title of this pull request.

Currently some test cases are commented out because it seems that there is no way to give predicate to SplitManager without deprecated version of `getSplits()` method. Actually, I'm not sure how I can port this part to new APIs.

martint commented on presto-jmx/src/test/java/io/prestosql/plugin/jmx/TestJmxSplitManager.java:
```diff
@@ -98,30 +98,30 @@ public void tearDown()
+//        for (Node node : nodes) {
+//            String nodeIdentifier = node.getNodeIdentifier();
+//
+//            // TODO Is there any way to give predicate to SplitManager without deprecated version of getSplits()?
```
The new API doesn't yet support pushing down predicates. It's something we're adding as part of https://github.com/prestosql/presto/issues/18. We'll add a version of the API that takes a `Constraint` object like `getTableLayout` does today to ease migration. We should probably wait to complete and merge this PR until that gets in (which will happen soon), otherwise, we're removing existing features for this connector.

takezoe replied:
Thanks for the explanation! I understand it wasn't a time to migrate connectors yet. I'm waiting for `Constraint` to be added.

martint replied:
@takezoe, with this, you should be able to complete the migration of this connector: https://github.com/prestosql/presto/pull/541.

martint commented:
@takezoe, https://github.com/prestosql/presto/pull/541 is merged, so you should be able to rebase this PR and implement the missing functionality now.

takezoe commented:
Thanks! 'm going to resume working on this pull request.

ebyhr commented on presto-jmx/src/main/java/io/prestosql/plugin/jmx/JmxTableHandle.java:
```diff
@@ -114,4 +126,60 @@ public ConnectorTableMetadata getTableMetadata()
+    public static class Node
+    {
+        private String identifier;
+        private HostAddress address;
```
I think these fields should be `private final`.

ebyhr commented on presto-jmx/src/test/java/io/prestosql/plugin/jmx/TestJmxMetadata.java:
```diff
@@ -96,4 +114,32 @@ public void testGetCumulativeTableHandle()
+        Optional<ConstraintApplicationResult<ConnectorTableHandle>> result = metadata.applyFilter(handle, new Constraint(TupleDomain.all()));
+
+        assertTrue(result.isPresent());
+        assertTrue(((JmxTableHandle) result.get().getHandle()).getNodes().size() == 3);
```
It would be better to use `assertEquals`. L138 is also.

ebyhr commented on presto-jmx/src/test/java/io/prestosql/plugin/jmx/TestJmxTableHandle.java:
```diff
@@ -48,23 +50,31 @@ public void testEquivalence()
+
+    private static JmxTableHandle.Node createTestingNode(String hostname)
+    {
+        return new JmxTableHandle.Node(hostname, HostAddress.fromString(format("%s:8080", hostname)));
```
`HostAddress.fromParts` is simpler than concating hostname and port.
```java
return new JmxTableHandle.Node(hostname, HostAddress.fromParts(hostname, 8080));
```

ebyhr commented on presto-jmx/src/main/java/io/prestosql/plugin/jmx/JmxSplitManager.java:
```diff
@@ -14,64 +14,29 @@
-                .map(node -> new JmxSplit(ImmutableList.of(node.getHostAndPort())))
-                .collect(toList());
+        List<ConnectorSplit> splits = tableHandle.getNodes().stream()
+                .map(node -> new JmxSplit(ImmutableList.of(node.getAddress()))).collect(toList());
```
Move `.collect(toList());` to next line.

takezoe replied:
> Move .collect(toList()); to next line.

Do we have a specific regulation about putting new lines? Actually, I have a similar code in `JmxMetadata` as follows. I'm wondering if I have to put new lines before `.map` and `.collect`.

```scala
private List<JmxTableHandle.Node> getAllNodes()
{
    return nodeManager.getAllNodes().stream().map(node -> new JmxTableHandle.Node(node.getNodeIdentifier(), node.getHostAndPort())).collect(toList());
}
```

ebyhr replied:
When the statement is simple, one line is fine. When you break a line after `.stream()`, subsequent each operations should be independent line as original code was. (There's no 'specific' regulation as far as I know).

The `getAllNodes` looks little long for one line and there's only one statement in the method, so putting new lines looks reasonable to me.

ebyhr commented on presto-jmx/src/main/java/io/prestosql/plugin/jmx/JmxRecordSetProvider.java:
```diff
@@ -153,7 +153,12 @@ else if (value.getClass().getComponentType() == short.class) {
+            ConnectorSession session,
+            ConnectorSplit split,
+            ConnectorTableHandle table,
+            List<? extends ColumnHandle> columns)
```
Unnecessary change for jmx migration. Basically, we separate the commit to fix styles.

takezoe replied:
I can't remember. Let me revert this change.

ebyhr commented on presto-jmx/src/main/java/io/prestosql/plugin/jmx/JmxTableHandle.java:
```diff
@@ -105,6 +116,7 @@ public String toString()
                 .add("objectNames", objectNames)
```
Found existing code has wrong combination in `toString()`.
```
.add("tableName", objectNames)
```

takezoe replied:
Nice catch! I will create another pull request to fix it.

takezoe replied:
Created: https://github.com/prestosql/presto/pull/634

ebyhr replied:
Thank you for your quick fix!

ebyhr commented on presto-jmx/src/main/java/io/prestosql/plugin/jmx/JmxMetadata.java:
```diff
@@ -242,17 +271,28 @@ public ColumnMetadata getColumnMetadata(ConnectorSession session, ConnectorTable
+                })
+                .collect(toList());
+
+        JmxTableHandle newTaleHandle = new JmxTableHandle(tableHandle.getTableName(), tableHandle.getObjectNames(), tableHandle.getColumnHandles(), tableHandle.isLiveData(), nodes);
```
Typo `newTaleHandle`

takezoe commented:
@martint I finished to migrate the JMX connector node selection pushdown to `Metadata.applyFilter`. Can you take a look?

martint commented on presto-jmx/src/main/java/io/prestosql/plugin/jmx/JmxTableHandle.java:
```diff
@@ -36,21 +36,21 @@
             @JsonProperty("columnHandles") List<JmxColumnHandle> columnHandles,
             @JsonProperty("liveData") boolean liveData,
-            @JsonProperty("addresses") List<HostAddress> addresses)
+            @JsonProperty("nodes") List<Node> nodes)
```
Why hold on to the list of nodes in the TableHandle? I'd be better to just hold on to the predicate and filter them out during split generation.

takezoe replied:
In my understanding, to hold the predicate ( `TupleDomain` or `Map<ColumnHandle, Domain>`) in the TableHandle, we need to make all classes held in the `ValueSet` serializable to JSON. Is this correct?

takezoe replied:
Oops, I didn't know `TupleDomain` is always JSON serializable. I fixed to hold `TupleDomain` in `TableHandle` instead of the list of node.

electrum commented on presto-jmx/src/main/java/io/prestosql/plugin/jmx/JmxMetadata.java:
```diff
@@ -96,6 +103,23 @@ public JmxTableHandle getTableHandle(ConnectorSession session, SchemaTableName t
+    @Override
+    public ConnectorTableProperties getTableProperties(ConnectorSession session, ConnectorTableHandle table)
+    {
+        return new ConnectorTableProperties(
```
This can use the no-args constructor, which does the same thing

martint commented:
@takezoe, can you rebase on master and remove the merge commits? They are confusing github and it's not displaying the actual differences with respect to the base commit.

takezoe commented:
Sure, I rebased on master branch without merge commits.

martint commented on presto-jmx/src/main/java/io/prestosql/plugin/jmx/JmxMetadata.java:
```diff
@@ -242,17 +254,14 @@ public ColumnMetadata getColumnMetadata(ConnectorSession session, ConnectorTable
-        return ImmutableList.of(new ConnectorTableLayoutResult(layout, constraint.getSummary()));
-    }
+        JmxTableHandle tableHandle = (JmxTableHandle) handle;
+        TupleDomain<ColumnHandle> predicate = constraint.getSummary();
```
`applyFilter` can get called multiple times throughout the optimization process. Connectors need to ensure that they "remember" the result of previous applications. Also, if applying the filter did not have any effect, the method needs to return `Optional.empty()`. Otherwise, the optimizer may loop forever.

See https://github.com/prestosql/presto/blob/master/presto-base-jdbc/src/main/java/io/prestosql/plugin/jdbc/JdbcMetadata.java#L86 for an example.

martint commented on presto-jmx/src/main/java/io/prestosql/plugin/jmx/JmxMetadata.java:
```diff
@@ -242,17 +254,14 @@ public ColumnMetadata getColumnMetadata(ConnectorSession session, ConnectorTable
-        return new ConnectorTableLayout(handle);
+        JmxTableHandle newTableHandle = new JmxTableHandle(tableHandle.getTableName(), tableHandle.getObjectNames(), tableHandle.getColumnHandles(), tableHandle.isLiveData(), predicate);
+
+        return Optional.of(new ConstraintApplicationResult(newTableHandle, predicate));
```
If the connector guarantees the results are going to satisfy the constraint, it's not necessary to return the predicate in the `ConstraintApplicationResult`. It can just indicate `TupleDomain.ALL`.

ebyhr commented on presto-jmx/src/main/java/io/prestosql/plugin/jmx/JmxTableHandle.java:
```diff
@@ -35,18 +37,21 @@
         this.objectNames = ImmutableList.copyOf(requireNonNull(objectNames, "objectName is null"));
         this.columnHandles = ImmutableList.copyOf(requireNonNull(columnHandles, "columnHandles is null"));
         this.liveData = liveData;
+        this.nodeFilter = requireNonNull(nodeFilter, "constraint is null");
```
I guess `nodeFilter is null` is expected message and L48 should be updated `@JsonProperty("nodeFilter")`.

takezoe replied:
Oh, that's true!

takezoe commented:
@martint Thanks for your review again. Fixed to return `Option.empty` if `JmxTableHandle` has a exact same constraint. Also now `JmxTableHandle` holds only node column filter and others are passed to `ConstraintApplicationResult` as `remainingFilter` to exclude predicates which are processed by the connector.

martint commented on presto-jmx/src/main/java/io/prestosql/plugin/jmx/JmxSplitManager.java:
```diff
@@ -52,22 +52,22 @@ public JmxSplitManager(NodeManager nodeManager)
-        TupleDomain<ColumnHandle> predicate = jmxLayout.getConstraint();
+        JmxTableHandle tableHandle = (JmxTableHandle) table;
 
-        //TODO is there a better way to get the node column?
```
unrelated change?

takezoe replied:
Oh, sorry. Reverted.

martint commented on presto-jmx/src/main/java/io/prestosql/plugin/jmx/JmxSplitManager.java:
```diff
@@ -52,22 +52,22 @@ public JmxSplitManager(NodeManager nodeManager)
         Optional<JmxColumnHandle> nodeColumnHandle = tableHandle.getColumnHandles().stream()
                 .filter(jmxColumnHandle -> jmxColumnHandle.getColumnName().equals(NODE_COLUMN_NAME))
                 .findFirst();
+
```
Unrelated formatting change.

takezoe replied:
True. Reverted too.

martint commented on presto-jmx/src/test/java/io/prestosql/plugin/jmx/TestJmxTableHandle.java:
```diff
@@ -48,23 +56,26 @@ public void testEquivalence()
                 .addEquivalentGroup(
-                        new JmxTableHandle(SCHEMA_TABLE_NAME, ImmutableList.of("name"), COLUMNS, false),
-                        new JmxTableHandle(SCHEMA_TABLE_NAME, ImmutableList.of("name"), COLUMNS, false))
+                        new JmxTableHandle(SCHEMA_TABLE_NAME, ImmutableList.of("nameX"), COLUMNS, true, TupleDomain.all()),
```
Why the name change here?

takezoe replied:
This is diff magic. I didn't change the name, just added new arguments. "name" patterns are coverd by preceding 2 test cases.

martint commented on presto-jmx/src/test/java/io/prestosql/plugin/jmx/TestJmxTableHandle.java:
```diff
@@ -48,23 +56,26 @@ public void testEquivalence()
+                        new JmxTableHandle(SCHEMA_TABLE_NAME, ImmutableList.of("nameX"), COLUMNS, false, TupleDomain.all()),
+                        new JmxTableHandle(SCHEMA_TABLE_NAME, ImmutableList.of("nameX"), COLUMNS, false, TupleDomain.all()))
                 .addEquivalentGroup(
-                        new JmxTableHandle(SCHEMA_TABLE_NAME, ImmutableList.of("nameX"), COLUMNS, false),
```
Why the name change here?

takezoe commented:
@martint Thanks for your comment! Fixed and squashed.

martint commented:
Merged. Thanks for your contribution!
